pyvrp/Model.py

```diff
@@ -0,0 +1,291 @@
+    ) -> Client:
+        """
+        Adds a client with the given attributes to the model. Returns the
+        created :class:`~pyvrp._ProblemData.Client` instance.
+        """
+        client = Client(
+            x,
+            y,
+            demand,
+            service_duration,
+            tw_early,
+            tw_late,
+            prize,
+            required,
+        )
+
+        self._clients.append(client)
+        return client
+
+    def add_depot(
+        self, x: int, y: int, tw_early: int = 0, tw_late: int = 0
+    ) -> Depot:
+        """
+        Adds a depot with the given attributes to the model. Returns the
+        created :class:`~pyvrp._ProblemData.Client` instance.
+
+        .. warning::
+
+           PyVRP does not yet support multi-depot VRPs. For now, only one depot
+           can be added to the model.
+        """
+        if len(self._depots) >= 1:
+            msg = "PyVRP does not yet support multi-depot VRPs."
+            raise ValueError(msg)
+
+        depot = Depot(x, y, tw_early=tw_early, tw_late=tw_late)
+        self._depots.append(depot)
+        return depot
+
+    def add_edge(
+        self,
+        frm: Union[Client, Depot],
+        to: Union[Client, Depot],
+        distance: int,
+        duration: int = 0,
+    ) -> Edge:
+        """
+        Adds an edge :math:`(i, j)` between ``frm`` (:math:`i`) and ``to``
+        (:math:`j`). The edge can be given distance and duration attributes.
+        Distance is required, but the default duration is zero. Returns the
+        created edge.
+
+        Raises
+        ------
+        ValueError
+            When either distance or duration is a negative value.
+        """
+        if distance < 0 or duration < 0:
+            raise ValueError("Cannot have negative edge distance or duration.")
+
+        edge = Edge(frm, to, distance, duration)
+        self._edges.append(edge)
+        return edge
+
+    def add_vehicle_type(self, number: int, capacity: int) -> VehicleType:
+        """
+        Adds a vehicle type with the given number of vehicles of given capacity
+        to the model. Returns the created vehicle type.
+
+        .. warning::
+
+           PyVRP does not yet support heterogeneous fleet VRPs. For now, only
+           one vehicle type can be added to the model.
+
+        Raises
+        ------
+        ValueError
+            When either the vehicle number or capacity is not a positive value.
+        """
+        if len(self._vehicle_types) >= 1:
+            msg = "PyVRP does not yet support heterogeneous fleet VRPs."
+            raise ValueError(msg)
+
+        if number <= 0:
+            raise ValueError("Must have positive number of vehicles.")
+
+        if capacity < 0:
+            raise ValueError("Cannot have negative vehicle capacity.")
+
+        vehicle_type = VehicleType(number, capacity)
+        self._vehicle_types.append(vehicle_type)
+        return vehicle_type
+
+    def update(self):
```

Comment:
Requiring the user to manually call `update()` and otherwise having outdated `data` feels a bit fragile to me. Can't we automatically invalidate `_data` and call `update()` when `.data` is accessed?